MAPL_Base/MAPL_GenericCplComp.F90

```diff
@@ -1103,4 +1122,437 @@ subroutine Finalize(CC, SRC, DST, CLOCK, RC)
+    integer, pointer                      :: mask(:) => null()
+    real, allocatable                     :: buf1(:), buf2(:,:), buf3(:,:,:)
+    real, pointer                         :: ptr1(:), ptr2(:,:), ptr3(:,:,:)
+
+! Begin...
+
+! Get the target components name and set-up traceback handle.
+! -----------------------------------------------------------
+
+    IAm = "MAPL_GenericCplComWriteRestart"
+    call ESMF_CplCompGet( CC, NAME=COMP_NAME, RC=STATUS )
+    _VERIFY(STATUS)
+    Iam = trim(COMP_NAME) // Iam
+
+! Retrieve the pointer to the internal state. It comes in a wrapper.
+! ------------------------------------------------------------------
+
+    call ESMF_CplCompGetInternalState ( CC, WRAP, STATUS )
+    _VERIFY(STATUS)
+
+    STATE     =>  WRAP%INTERNAL_STATE
+
+    call ESMF_CplCompGet(CC, vm=vm, rc=status)
+    _VERIFY(STATUS)
+
+    filename = trim(state%name) // '_chk' ! following Andrea's suggestion
+    am_i_root = MAPL_AM_I_ROOT(vm)
+
+    unit=0 ! just to initialize
+    n_vars = size(state%src_spec)
+    if (am_i_root) then
+       UNIT = GETFILE(filename, rc=status)
+       _VERIFY(status)
+       write(unit) n_vars
+    end if
+
+       ! for each var
+    do i = 1, n_vars
+       ! varname we can get from query SHORT_NAME in state%src_spec(i)
+       call MAPL_VarSpecGet(state%src_spec(i), SHORT_NAME=name, rc=status)
+       _VERIFY(status)
+       call ESMF_StateGet(SRC, name, field=field, rc=status) 
+       _VERIFY(status)
+       call ESMF_FieldGet(field, grid=grid, rc=status)
+       _VERIFY(status)
+
+       rank = state%accum_rank(i)
+       call ESMF_AttributeGet(field, name='DIMS', value=DIMS, rc=status)
+       _VERIFY(STATUS)
+       mask => null()
+       if (DIMS == MAPL_DimsTileOnly .or. DIMS == MAPL_DimsTileTile) then
+          call MAPL_TileMaskGet(grid,  mask, rc=status)
+          _VERIFY(STATUS)
+       end if
+       ! ALT note: calling a procedure with optional argument, and passing NULL pointer to indicate "absent", needs ifort16 or newer
+
+       !we need to get the MAX n_count         
+       call MAPL_CommsAllReduceMax(vm, sendbuf=state%accum_count(i), &
+            recvbuf=n_count, cnt=1, RC=status)
+       _VERIFY(status)
+       if (am_i_root) then
+          write(unit) n_count
+       end if
+       select case (rank)
+          case(1)
+             local_undefs = associated(state%array_count(i)%ptr1c)
+          case(2)
+             local_undefs = associated(state%array_count(i)%ptr2c)
+          case(3)
+             local_undefs = associated(state%array_count(i)%ptr3c)
+          case default
+             ASSERT_(.false.)
+          end select
+       have_undefs = 0
+       n_undefs = 0
+       if (local_undefs) have_undefs = 1
+       call MAPL_CommsAllReduceMax(vm, sendbuf=have_undefs, &
+            recvbuf=n_undefs, cnt=1, RC=status)
+       _VERIFY(status)
+       if (am_i_root) then
+          write(unit) n_undefs
+       end if
+
+       select case(rank)
```

Comment:
Before you know it, we will need rank 4 as well (ala Arlindo's recent issue in GOCART).   I think you should unfortunately consider a #include type of approach that makes the per-rank logic generic.  Ugh.

Reply:
Note that this comment applies to the previous procedure (ReadRestart) as well.    Even if you decide against the #include approach, I think you should consider using internal procedures to reduce the size/complexity of these main procedures.

Reply:
I will take a closer look. Supporting 4-d in the coupler is straight forward but tedious. If we first do the split of the 4d fields before we come here, this might not be necessary